### PR TITLE
Fixes RuntimeError-No package.json issue

### PR DESCRIPTION
**Pull request by sachin-sandhu**
### What are you trying to accomplish?

**Preface:** Fixes issue related with `RuntimeError-No package.json` and `Dependabot::SharedHelpers::HelperSubprocessFailed`

**Fixes:** 

1. Fixes user runtime exception ` RuntimeError::No package.json!` with `DependencyFileNotFound` exception.
2. Allows exception handling for 
 a. `Unauthenticated: request did not include an Authorization header.`
 b. `ESOCKETTIMEDOUT`
 c. `socket hang up`

### Checklist

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

### CODE REVIEW

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb [hunk 1]
```diff
@@ -166,7 +166,7 @@ def build_dependency(file:, type:, name:, requirement:)
 
       sig { override.void }
       def check_required_files
-        raise "No package.json!" unless get_original_file("package.json")
+        raise DependencyFileNotFound.new(nil, "package.json not found.") unless get_original_file("package.json")
```
Comment (abdulapopoola) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:169]: can we please add unit tests to capture this failure mode?

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater.rb [hunk 2]
```diff
@@ -139,7 +139,7 @@ def filtered_dependency_files
 
       sig { override.void }
       def check_required_files
-        raise "No package.json!" unless get_original_file("package.json")
+        raise DependencyFileNotFound.new(nil, "package.json not found.") unless get_original_file("package.json")
```
Comment (abdulapopoola) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater.rb:142]: can we please add unit tests to capture this failure mode?

File: npm_and_yarn/lib/dependabot/npm_and_yarn.rb [hunk 3]
```diff
@@ -56,6 +56,10 @@ module NpmAndYarn
     # Used to check if error message contains timeout fetching package
     TIMEOUT_FETCHING_PACKAGE_REGEX = %r{(?<url>.+)/(?<package>[^/]+): ETIMEDOUT}
 
+    ESOCKETTIMEDOUT = /(?<package>.*?): ESOCKETTIMEDOUT/
+
+    SOCKET_HANG_UP = /(?<url>.*?): socket hang up/
```
Comment (abdulapopoola) [inline npm_and_yarn/lib/dependabot/npm_and_yarn.rb:61]: where are these used?

File: npm_and_yarn/lib/dependabot/npm_and_yarn.rb [hunk 4]
```diff
@@ -56,6 +56,10 @@ module NpmAndYarn
     # Used to check if error message contains timeout fetching package
     TIMEOUT_FETCHING_PACKAGE_REGEX = %r{(?<url>.+)/(?<package>[^/]+): ETIMEDOUT}
 
+    ESOCKETTIMEDOUT = /(?<package>.*?): ESOCKETTIMEDOUT/
```
Comment (abdulapopoola) [inline npm_and_yarn/lib/dependabot/npm_and_yarn.rb:59]: where is this error used?

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb (hunk 1 above)
Comment (sachin-sandhu) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:169]: @abdulapopoola , added test cases

File: npm_and_yarn/lib/dependabot/npm_and_yarn.rb (hunk 4 above)
Comment (sachin-sandhu) [reply npm_and_yarn/lib/dependabot/npm_and_yarn.rb:59]: @abdulapopoola , i have updated rest of code.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater.rb (hunk 2 above)
Comment (sachin-sandhu) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater.rb:142]: @abdulapopoola , added test cases.

**Review by abdulapopoola — APPROVED** [review]

**Pull request merged**